File: backend/internal/adapters/http/middleware/roles.go
package middleware

import (
	"net/http"

	"github.com/careplus/pharmacy-backend/internal/adapters/http/dto/response"
	"github.com/careplus/pharmacy-backend/pkg/errors"
	"github.com/gin-gonic/gin"
)

// Role constants used across middleware and handlers.
const (
	RoleAdmin      = "admin"
	RoleManager    = "manager"
	RolePharmacist = "pharmacist"
	RoleStaff      = "staff" // end-user / buyer
)

// StaffRoles are roles that can access management features (excludes end-user "staff").
var StaffRoles = []string{RoleAdmin, RoleManager, RolePharmacist}

// RequireAnyRole returns a middleware that allows only the given roles.
// Use after Auth middleware. Returns 403 if the user's role is not in the list.
func RequireAnyRole(allowedRoles ...string) gin.HandlerFunc {
	set := make(map[string]bool)
	for _, r := range allowedRoles {
		set[r] = true
	}
	return func(c *gin.Context) {
		role, exists := c.Get("role")
		if !exists {
			c.JSON(http.StatusForbidden, response.ErrorResponse{Code: errors.ErrCodeForbidden, Message: "role not set"})
			c.Abort()
			return
		}
		roleStr, ok := role.(string)
		if !ok || !set[roleStr] {
			c.JSON(http.StatusForbidden, response.ErrorResponse{Code: errors.ErrCodeForbidden, Message: "insufficient role for this action"})
			c.Abort()
			return
		}
		c.Next()
	}
}

// RequireStaffRole allows admin, manager, or pharmacist (not end-user "staff").
// Use for product/category/inventory/invoice/payment management and similar.
func RequireStaffRole() gin.HandlerFunc {
	return RequireAnyRole(StaffRoles...)
}

File: frontend/src/lib/roles.ts
/**
 * Role constants aligned with backend. Used for nav visibility and route guards.
 */
export const ROLE_ADMIN = 'admin';
export const ROLE_MANAGER = 'manager';
export const ROLE_PHARMACIST = 'pharmacist';
export const ROLE_STAFF = 'staff'; // end-user / buyer

/** Roles that see the full staff dashboard (product/order management, inventory, etc.). */
export const STAFF_DASHBOARD_ROLES: string[] = [ROLE_ADMIN, ROLE_MANAGER, ROLE_PHARMACIST];

/** All roles that can access the app dashboard (including buyer). */
export const DASHBOARD_ROLES: string[] = [...STAFF_DASHBOARD_ROLES, ROLE_STAFF];

/** Paths that end-users (role staff) are allowed to access. Others redirect to /dashboard. */
export const BUYER_ALLOWED_PATHS = ['/dashboard', '/catalog', '/orders', '/chat', '/profile'];

export function isStaffRole(role: string | undefined): boolean {
  return !!role && STAFF_DASHBOARD_ROLES.includes(role);
}

export function isBuyerRole(role: string | undefined): boolean {
  return role === ROLE_STAFF;
}

export function isBuyerAllowedPath(pathname: string): boolean {
  return BUYER_ALLOWED_PATHS.some((p) => pathname === p || pathname.startsWith(p + '/'));
}
